libs/full/collectives/include/hpx/collectives/detail/communication_set_node.hpp

```diff
@@ -0,0 +1,172 @@
+//  Copyright (c) 2020 Hartmut Kaiser
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+
+#if !defined(HPX_COMPUTE_DEVICE_CODE)
+
+#include <hpx/actions_base/component_action.hpp>
+#include <hpx/assert.hpp>
+#include <hpx/datastructures/any.hpp>
+#include <hpx/lcos_local/and_gate.hpp>
+#include <hpx/modules/futures.hpp>
+#include <hpx/runtime/components/server/component_base.hpp>
+#include <hpx/runtime/naming/id_type.hpp>
+#include <hpx/synchronization/spinlock.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace traits {
+
+    // This type will be specialized for a particular collective operation
+    template <typename Communicator, typename Operation>
+    struct communication_operation;
+
+}}    // namespace hpx::traits
+
+namespace hpx { namespace lcos { namespace detail {
+
+    ///////////////////////////////////////////////////////////////////////////
+    HPX_EXPORT std::size_t calculate_connected_node(
+        std::size_t site, std::size_t arity);
+    HPX_EXPORT std::size_t calculate_num_connected(
+        std::size_t num_sites, std::size_t site, std::size_t arity);
+
+    ///////////////////////////////////////////////////////////////////////////
+    constexpr std::size_t next_power_of_two(std::uint64_t v)
+    {
+        --v;
+        v |= v >> 1;
+        v |= v >> 2;
+        v |= v >> 4;
+        v |= v >> 8;
+        v |= v >> 16;
+        v |= v >> 32;
+        return static_cast<std::size_t>(++v);
```

Comment:
Don't know if this is where you got this from, but I just have to link it as it's a very useful resource for these kind of things: https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2.